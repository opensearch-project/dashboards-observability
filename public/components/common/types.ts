/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export interface FieldCapAttributes {
  type: string;
  searchable: boolean;
  aggregatable: boolean;
}

export interface FieldCapResponse {
  indices: string[];
  fields: Record<string, Record<string, FieldCapAttributes>>;
}
