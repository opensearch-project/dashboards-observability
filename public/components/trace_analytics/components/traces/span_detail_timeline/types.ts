/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Span } from '../../common/constants';

export interface SpanWithChildren extends Span {
  children: SpanWithChildren[];
}

export interface SpanReference {
  refType: 'CHILD_OF' | 'FOLLOWS_FROM';
  spanID: string;
}

export type JaegerSpan = Omit<Span, 'spanId'> & {
  spanID: string;
  duration: number;
  references: SpanReference[];
};

export type SpanMap = Record<string, SpanWithChildren>;
