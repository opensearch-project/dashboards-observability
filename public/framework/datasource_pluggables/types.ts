/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { IDataFetcher } from '../../services/data_fetchers/fetch_interface';

export interface IDataSourceComponentSet {
  ui: {
    QueryEditor: React.ReactNode;
    ConfigEditor: React.ReactNode;
    SidePanel: React.ReactNode;
  };
  services: {
    data_fetcher: IDataFetcher;
  };
}

export interface IDataSourcePluggableComponents {
  languages?: Record<string, IDataSourceComponentSet>;
  // Other variation keys can be added in the future
}
