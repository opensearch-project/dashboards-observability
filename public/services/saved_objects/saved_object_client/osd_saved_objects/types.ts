/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { SavedObjectAttributes, SimpleSavedObject } from '../../../../../../../src/core/public';
import { OBSERVABILTY_SAVED_OBJECTS } from '../../../../../common/types/observability_saved_object_attributes';
import { SavedObjectsCreateResponse } from '../types';

export type ObservabilitySavedObjectsType = typeof OBSERVABILTY_SAVED_OBJECTS[number];

export interface OSDSavedObjectCreateResponse<T extends SavedObjectAttributes>
  extends SavedObjectsCreateResponse {
  object: SimpleSavedObject<T>;
}

export interface OSDSavedObjectUpdateResponse<T extends SavedObjectAttributes>
  extends SavedObjectsCreateResponse {
  object: SimpleSavedObject<Partial<T>>;
}
