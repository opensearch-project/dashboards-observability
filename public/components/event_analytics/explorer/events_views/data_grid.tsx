/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  EuiDataGrid,
  EuiDataGridColumn,
  EuiDataGridSorting,
  EuiDescriptionList,
  EuiDescriptionListDescription,
  EuiDescriptionListTitle,
  EuiPanel,
  EuiDataGridProps,
} from '@elastic/eui';
import moment from 'moment';
import React, { Fragment, MutableRefObject, useEffect, useRef, useState } from 'react';
import { HttpSetup } from '../../../../../../../src/core/public';
import {
  DATE_DISPLAY_FORMAT,
  DEFAULT_EMPTY_EXPLORER_FIELDS,
  DEFAULT_SOURCE_COLUMN,
  DEFAULT_TIMESTAMP_COLUMN,
} from '../../../../../common/constants/explorer';
import { IExplorerFields, IField } from '../../../../../common/types/explorer';
import PPLService from '../../../../services/requests/ppl';
import { useFetchEvents } from '../../hooks';
import { redoQuery } from '../../utils/utils';
import { FlyoutButton } from './docViewRow';

export interface DataGridProps {
  http: HttpSetup;
  pplService: PPLService;
  rows: any[];
  explorerFields: IExplorerFields;
  timeStampField: string;
  rawQuery: string;
  totalHits: number;
  requestParams: any;
  startTime: string;
  endTime: string;
  storedSelectedColumns: IField[];
  formatGridColumn?: (columns: EuiDataGridColumn[]) => EuiDataGridColumn[];
  OuiDataGridProps?: Partial<EuiDataGridProps>;
}

const defaultFormatGrid = (columns: EuiDataGridColumn[]) => columns;

export function DataGrid(props: DataGridProps) {
  const {
    http,
    pplService,
    rows,
    explorerFields,
    timeStampField,
    rawQuery,
    totalHits,
    requestParams,
    startTime,
    endTime,
    formatGridColumn = defaultFormatGrid,
    OuiDataGridProps,
  } = props;
  const { fetchEvents } = useFetchEvents({
    pplService,
    requestParams,
  });
  const selectedColumns =
    explorerFields.selectedFields.length > 0
      ? explorerFields.selectedFields
      : DEFAULT_EMPTY_EXPLORER_FIELDS;
  // useRef instead of useState somehow solves the issue of user triggered sorting not
  // having any delays
  const sortingFields: MutableRefObject<EuiDataGridSorting['columns']> = useRef([]);
  const pageFields = useRef([0, 100]);

  const [data, setData] = useState(rows);

  useEffect(() => {
    setData(rows);
  }, [rows]);

  // setSort and setPage are used to change the query and send a direct request to get data
  const setSort = (sort: EuiDataGridSorting['columns']) => {
    sortingFields.current = sort;

    redoQuery(
      startTime,
      endTime,
      rawQuery,
      timeStampField,
      sortingFields,
      pageFields,
      fetchEvents,
      setData
    );
  };

  const setPage = (page: number[]) => {
    pageFields.current = page;
    redoQuery(
      startTime,
      endTime,
      rawQuery,
      timeStampField,
      sortingFields,
      pageFields,
      fetchEvents,
      setData
    );
  };

  // creates the header for each column listing what that column is
  const dataGridColumns = () => {
    const columns: EuiDataGridColumn[] = [];
    selectedColumns.map(({ name }) => {
      if (name === 'timestamp') {
        columns.push(DEFAULT_TIMESTAMP_COLUMN);
      } else if (name === '_source') {
        columns.push(DEFAULT_SOURCE_COLUMN);
      } else {
        columns.push({
          id: name,
          display: name,
          isSortable: true, // TODO: add functionality here based on type
        });
      }
    });
    return formatGridColumn(columns);
  };

  // used for which columns are visible and their order
  const dataGridColumnVisibility = () => {
    if (selectedColumns.length > 0) {
      const columns: string[] = [];
      selectedColumns.map(({ name }) => {
        columns.push(name);
      });
      return {
        visibleColumns: columns,
        setVisibleColumns: () => {
          // TODO: implement with sidebar field order (dragability) changes
        },
      };
    }
    // default shown fields
    throw new Error('explorer data grid stored columns empty');
  };

  // sets the very first column, which is the button used for the flyout of each row
  const dataGridLeadingColumns = () => {
    return [
      {
        id: 'inspectCollapseColumn',
        headerCellRender: () => null,
        rowCellRender: ({ rowIndex }: { rowIndex: number }) => {
          return (
            <FlyoutButton
              ref={null}
              http={http}
              key={null}
              docId={'undefined'}
              doc={rows[rowIndex % pageFields.current[1]]}
              selectedCols={explorerFields.queriedFields}
              timeStampField={timeStampField}
              explorerFields={explorerFields}
              pplService={pplService}
              rawQuery={rawQuery}
              onFlyoutOpen={() => {}}
              dataGridColumns={dataGridColumns}
              dataGridColumnVisibility={dataGridColumnVisibility}
              selectedIndex={rowIndex}
              sortingFields={sortingFields}
              rowHeightsOptions={rowHeightsOptions}
              rows={rows}
            />
          );
        },
        width: 40,
      },
    ];
  };

  // renders what is shown in each cell, i.e. the content of each row
  const dataGridCellRender = ({ rowIndex, columnId }: { rowIndex: number; columnId: string }) => {
    const trueIndex = rowIndex % pageFields.current[1];
    if (trueIndex < data.length) {
      if (columnId === '_source') {
        return (
          <EuiDescriptionList type="inline" compressed>
            {Object.keys(data[trueIndex]).map((key) => (
              <Fragment key={key}>
                <EuiDescriptionListTitle className="osdDescriptionListFieldTitle">
                  {key}
                </EuiDescriptionListTitle>
                <EuiDescriptionListDescription>
                  {data[trueIndex][key]}
                </EuiDescriptionListDescription>
              </Fragment>
            ))}
          </EuiDescriptionList>
        );
      }
      if (columnId === 'timestamp') {
        return `${moment(data[trueIndex][columnId]).format(DATE_DISPLAY_FORMAT)}`;
      }
      return `${data[trueIndex][columnId]}`;
    }
    return null;
  };

  // ** Pagination config
  const [pagination, setPagination] = useState({ pageIndex: 0, pageSize: 100 });
  // changing the number of items per page, reset index and modify page size
  const onChangeItemsPerPage = (pageSize) =>
    setPagination(() => {
      setPage([0, pageSize]);
      return { pageIndex: 0, pageSize };
    });
  // changing the page index, keep page size constant
  const onChangePage = (pageIndex) => {
    setPagination(({ pageSize }) => {
      setPage([pageIndex, pageSize]);
      return { pageSize, pageIndex };
    });
  };

  const rowHeightsOptions = () => ({
    defaultHeight: {
      // if source is listed as a column, add extra space
      lineCount: selectedColumns.some((obj) => obj.name === '_source') ? 3 : 1,
    },
  });

  // TODO: memoize the expensive table below

  return (
    <EuiPanel paddingSize="s">
      <div className="dscTable dscTableFixedScroll">
        <EuiDataGrid
          aria-labelledby="aria-labelledby"
          data-test-subj="docTable"
          columns={dataGridColumns()}
          columnVisibility={dataGridColumnVisibility()}
          leadingControlColumns={dataGridLeadingColumns()}
          rowCount={totalHits}
          renderCellValue={dataGridCellRender}
          pagination={{
            ...pagination,
            pageSizeOptions: [25, 50, 100],
            onChangePage,
            onChangeItemsPerPage,
          }}
          sorting={{
            columns: sortingFields.current,
            onSort: setSort,
          }}
          toolbarVisibility={{
            showColumnSelector: {
              allowHide: false,
              allowReorder: true,
            },
            showFullScreenSelector: false,
            showStyleSelector: false,
          }}
          rowHeightsOptions={rowHeightsOptions()}
          {...OuiDataGridProps}
        />
      </div>
    </EuiPanel>
  );
}
