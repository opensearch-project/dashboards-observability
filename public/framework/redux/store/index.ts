/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configureStore } from '@reduxjs/toolkit';
import storage from 'redux-persist/lib/storage/session';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import { rootReducer } from '../reducers';

const persistConfig = {
  key: 'root',
  storage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const dataStore = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }),
  devTools: process.env.NODE_ENV !== 'production',
  enhancers: [],
});

if (process.env.NODE_ENV === 'development' && module.hot) {
  module.hot.accept('./rootReducer', () => {
    store.replaceReducer(persistReducer(persistConfig, require('./rootReducer').default)); // eslint-disable-line @typescript-eslint/no-var-requires
  });
}

export type AppDispatch = typeof dataStore.dispatch;

export const persistor = persistStore(dataStore);

export const store = dataStore;
