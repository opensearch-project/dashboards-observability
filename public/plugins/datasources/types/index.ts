/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export {
  IDatasourcePickerProps,
  IDatasourceListOption,
  IDatasourcePickerUIStyles,
  IDatasourceSelectableOption,
  ISelectedSourceOption,
} from './datasource_picker';
