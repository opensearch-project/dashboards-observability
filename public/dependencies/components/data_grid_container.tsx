/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState, useEffect, useCallback } from 'react';
import { useSelector } from 'react-redux';
import { EuiDataGridColumn } from '@elastic/eui';
import {
  DataGrid,
  DataGridProps,
} from '../../components/event_analytics/explorer/events_views/data_grid';
import { getOSDHttp, getPPLService } from '../../../common/utils';
import { selectFields } from '../../components/event_analytics/redux/slices/field_slice';
import { initialTabId } from '../../framework/redux/store/shared_state';
import { useFetchEvents } from '../../components/event_analytics/hooks';
import { redoQuery } from '../../components/event_analytics/utils/utils';
import { DEFAULT_TIMESTAMP_COLUMN } from '../../../common/constants/explorer';
import { RenderProps } from '../../types';

export const DataGridContainer = (
  props: Pick<DataGridProps, 'rawQuery'> & { renderProps: RenderProps }
) => {
  const { renderProps, ...others } = props;
  const pplService = getPPLService();
  const http = getOSDHttp();
  const tabId = initialTabId;
  const explorerFields = useSelector(selectFields)[tabId];
  const { fetchEvents } = useFetchEvents({
    pplService,
    requestParams: {
      tabId,
    },
  });
  const [data, setData] = useState<any[]>([]);
  const isFullScreen = renderProps.chatContext.flyoutFullScreen;
  useEffect(() => {
    if (props.rawQuery) {
      redoQuery(
        '',
        '',
        props.rawQuery || '',
        '',
        { current: [] },
        { current: [0, 5] },
        fetchEvents,
        setData
      );
    }
  }, [props.rawQuery]);
  const formatGridColumn = useCallback(
    (columns: EuiDataGridColumn[]) =>
      columns.map((item) => {
        if (item.id === DEFAULT_TIMESTAMP_COLUMN.id) {
          const { initialWidth, ...otherConfig } = item;
          return {
            ...otherConfig,
            initialWidth: renderProps.chatContext.flyoutFullScreen ? initialWidth : 50,
          };
        }

        return item;
      }),
    [renderProps.chatContext.flyoutFullScreen]
  );
  if (!data.length || !props.rawQuery) {
    return null;
  }
  return (
    <div style={{ overflowX: 'hidden' }}>
      <DataGrid
        http={http}
        pplService={pplService}
        rows={data}
        explorerFields={explorerFields}
        timeStampField="timestamp"
        totalHits={data.length}
        requestParams={{
          tabId,
        }}
        startTime={''}
        endTime="now"
        storedSelectedColumns={[]}
        formatGridColumn={formatGridColumn}
        OuiDataGridProps={{
          gridStyle: isFullScreen
            ? {}
            : {
                fontSize: 's',
                cellPadding: 's',
              },
          /**
           * Chatbot's default width is smaller than 478
           */
          minSizeForControls: 300,
          /**
           * In non-fullscreen mode, it requires at least 80px to avoid truncate.
           */
          rowHeightsOptions: {
            defaultHeight: isFullScreen ? undefined : 80,
          },
        }}
        {...others}
      />
    </div>
  );
};
