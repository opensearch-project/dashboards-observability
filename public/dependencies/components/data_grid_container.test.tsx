/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { DataGridContainer } from './data_grid_container';
import { RenderProps } from '../../types';
import { store } from '../../framework/redux/store';
import { DataGridProps } from '../../components/event_analytics/explorer/events_views/data_grid';
import {
  DEFAULT_SOURCE_COLUMN as mock_DEFAULT_SOURCE_COLUMN,
  DEFAULT_TIMESTAMP_COLUMN as mock_DEFAULT_TIMESTAMP_COLUMN,
} from '../../../common/constants/explorer';

jest.mock('../../components/event_analytics/utils/utils', () => ({
  redoQuery: (...args: any[]) => {
    args[7]([
      {
        timestamp: 'now',
        count: 1,
      },
    ]);
  },
}));

jest.mock('../../../common/utils', () => ({
  getPPLService: () => ({}),
  getOSDHttp: () => ({}),
}));

jest.mock('../../components/event_analytics/explorer/events_views/data_grid', () => ({
  DataGrid: (props: DataGridProps) => {
    const columnsResult = props.formatGridColumn?.([
      mock_DEFAULT_TIMESTAMP_COLUMN,
      mock_DEFAULT_SOURCE_COLUMN,
    ]);
    return (
      <div data-test-subj="test">
        {JSON.stringify({ ...props, columns: columnsResult }, null, 2)}
      </div>
    );
  },
}));

describe('<DataGridContainer />', () => {
  it('should render when all props are provided', async () => {
    const { container, findByTestId } = render(
      <Provider store={store}>
        <DataGridContainer
          rawQuery="source=foo"
          renderProps={{
            props: {
              message: {
                type: 'output',
                contentType: 'ppl_data_grid',
                content: 'source=foo',
              },
            },
            chatContext: {
              flyoutFullScreen: false,
            } as RenderProps['chatContext'],
          }}
        />
      </Provider>
    );
    await findByTestId('test');
    expect(container).toMatchSnapshot();
  });

  it('should render empty when rawQuery is not provided', async () => {
    const { container, queryByTestId } = render(
      <Provider store={store}>
        <DataGridContainer
          rawQuery=""
          renderProps={{
            props: {
              message: {
                type: 'output',
                contentType: 'ppl_data_grid',
                content: 'source=foo',
              },
            },
            chatContext: {
              flyoutFullScreen: false,
            } as RenderProps['chatContext'],
          }}
        />
      </Provider>
    );
    await waitFor(() => {
      expect(queryByTestId('test')).toBeNull();
    });
    expect(container).toMatchSnapshot();
  });

  it('should switch props when in full screen mode', async () => {
    const { container, findByTestId } = render(
      <Provider store={store}>
        <DataGridContainer
          rawQuery="source=foo"
          renderProps={{
            props: {
              message: {
                type: 'output',
                contentType: 'ppl_data_grid',
                content: 'source=foo',
              },
            },
            chatContext: {
              flyoutFullScreen: true,
            } as RenderProps['chatContext'],
          }}
        />
      </Provider>
    );
    await findByTestId('test');
    expect(container).toMatchSnapshot();
  });
});
