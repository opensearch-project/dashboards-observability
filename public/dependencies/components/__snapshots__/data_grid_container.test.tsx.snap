// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DataGridContainer /> should render empty when rawQuery is not provided 1`] = `<div />`;

exports[`<DataGridContainer /> should render when all props are provided 1`] = `
<div>
  <div
    style="overflow-x: hidden;"
  >
    <div
      data-test-subj="test"
    >
      {
  "http": {},
  "pplService": {},
  "rows": [
    {
      "timestamp": "now",
      "count": 1
    }
  ],
  "explorerFields": {
    "selectedFields": [],
    "unselectedFields": [],
    "availableFields": [],
    "queriedFields": []
  },
  "timeStampField": "timestamp",
  "totalHits": 1,
  "requestParams": {
    "tabId": "OBSERVABILITY_DEFAULT_TAB"
  },
  "startTime": "",
  "endTime": "now",
  "storedSelectedColumns": [],
  "OuiDataGridProps": {
    "gridStyle": {
      "fontSize": "s",
      "cellPadding": "s"
    },
    "minSizeForControls": 300,
    "rowHeightsOptions": {
      "defaultHeight": 80
    }
  },
  "rawQuery": "source=foo",
  "columns": [
    {
      "id": "timestamp",
      "isSortable": true,
      "display": "Time",
      "schema": "datetime",
      "initialWidth": 50
    },
    {
      "id": "_source",
      "isSortable": false,
      "display": "Source",
      "schema": "_source"
    }
  ]
}
    </div>
  </div>
</div>
`;

exports[`<DataGridContainer /> should switch props when in full screen mode 1`] = `
<div>
  <div
    style="overflow-x: hidden;"
  >
    <div
      data-test-subj="test"
    >
      {
  "http": {},
  "pplService": {},
  "rows": [
    {
      "timestamp": "now",
      "count": 1
    }
  ],
  "explorerFields": {
    "selectedFields": [],
    "unselectedFields": [],
    "availableFields": [],
    "queriedFields": []
  },
  "timeStampField": "timestamp",
  "totalHits": 1,
  "requestParams": {
    "tabId": "OBSERVABILITY_DEFAULT_TAB"
  },
  "startTime": "",
  "endTime": "now",
  "storedSelectedColumns": [],
  "OuiDataGridProps": {
    "gridStyle": {},
    "minSizeForControls": 300,
    "rowHeightsOptions": {}
  },
  "rawQuery": "source=foo",
  "columns": [
    {
      "id": "timestamp",
      "isSortable": true,
      "display": "Time",
      "schema": "datetime",
      "initialWidth": 200
    },
    {
      "id": "_source",
      "isSortable": false,
      "display": "Source",
      "schema": "_source"
    }
  ]
}
    </div>
  </div>
</div>
`;
